File: Sliding Window & Two Pointer/TrappingRainwater.cpp
// 42. Trapping Rain Water

// Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.

// Using extra space for prefixSum and suffixSum Arrays => O(2n) space complexity
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> prefixMax(n), suffixMax(n);

        prefixMax[0] = height[0];
        for (int i = 1; i < n; i++) {
            prefixMax[i] = max(height[i], prefixMax[i - 1]);
        }

        suffixMax[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            suffixMax[i] = max(suffixMax[i + 1], height[i]);
        }

        int total = 0;
        for (int i = 0; i < n; i++) {
            total += min(prefixMax[i], suffixMax[i]) - height[i];
        }
        return total;
    }
};

// Without any extra space => O(1) space complexity
class Solution2 {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int leftMax = 0, rightMax = 0, total = 0;
        int low = 0, high = n - 1;
      
        while (low < high) {
            if (height[low] <= height[high]) {
                if (height[low] < leftMax) {
                    total += leftMax - height[low];
                } else {
                    leftMax = height[low];
                }
                low++;
            } else {
                if (height[high] < rightMax) {
                    total += rightMax - height[high];
                } else {
                    rightMax = height[high];
                }
                high--;
            }
        }
        return total;
    }
};

File: Sliding Window & Two Pointer/TrappingRainwater.py
# https://leetcode.com/problems/trapping-rain-water/
from typing import List

# O(2n) -> Space Complexity
class Solution1:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if n == 0: return 0
        
        prefixMax: List[int] = [height[0]] * n
        for i in range(1, n):
            prefixMax[i] = max(prefixMax[i-1], height[i])
            
        suffixMax: List[int] = [height[n-1]] * n
        for i in range(n-2, -1, -1):
            suffixMax[i] = max(suffixMax[i+1], height[i])
            
        res = 0
        for i in range(0, n):
            res += min(prefixMax[i], suffixMax[i]) - height[i]
        return res
    
# O(1) -> Space Complexity
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l, r = 0, n-1
        leftMax, rightMax, res = 0, 0, 0
        while l < r:
            if height[l] <= height[r]:
                if leftMax > height[l]:
                    res += leftMax - height[l]
                else: 
                    leftMax = height[l]
                l += 1
            else:
                if rightMax > height[r]:
                    res += rightMax - height[r]
                else:
                    rightMax = height[r]
                r -= 1
        return res
